feat(apple): Migration Guide for V9

philipphofmann opened the pull request:
<!-- Use this checklist to make sure your PR is ready for merge. You may delete any sections you don't need. -->

## DESCRIBE YOUR PR

Add the migration guide for major version v9. We still have to wait for Cocoa V9 to be shipped on Monday Dec 1st to merge this PR.

We also need to double check if we merged https://github.com/getsentry/sentry-cocoa/issues/6327 before the release, if yes, we must include it in here.

## IS YOUR CHANGE URGENT?  

Help us prioritize incoming PRs by letting us know when the change needs to go live.
- [ ] Urgent deadline (GA date, etc.): <!-- ENTER DATE HERE -->
- [x] Other deadline: Ideally, Monday, Dec 1st.
- [ ] None: Not urgent, can wait up to 1 week+

## SLA

- Teamwork makes the dream work, so please add a reviewer to your PRs.
- Please give the docs team up to 1 week to review your PR unless you've added an urgent due date to it.
Thanks in advance for your help!

## PRE-MERGE CHECKLIST

*Make sure you've checked the following before merging your changes:*

- [x] Checked Vercel preview for correctness, including links
- [ ] PR was reviewed and approved by any necessary SMEs (subject matter experts)
- [ ] PR was reviewed and approved by a member of the [Sentry docs team](https://github.com/orgs/getsentry/teams/docs)

philprime commented on docs/platforms/apple/common/migration/index.mdx:
```diff
@@ -3,7 +3,126 @@ title: Migration Guide
+
+We now enabled the following features by default:
+
+- **Performance V2 is enabled by default**: The app start duration now finishes when the first frame is drawn instead of when the OS posts the `UIWindowDidBecomeVisibleNotification`. We removed the `enablePerformanceV2` option.
```
This is valid feedback

philprime commented on docs/platforms/apple/common/migration/index.mdx:
```diff
@@ -3,7 +3,126 @@ title: Migration Guide
+#### Option Changes
+
+- `enableFileManagerSwizzling` is now a top-level non-experimental option and remains disabled by default.
+- `enableDataSwizzling` is now a top-level option and is enabled by default.
```
They're still shown as experimental.enableFileManagerSwizzling and experimental.enableDataSwizzling in code examples.

File affected:
- docs/platforms/apple/common/tracing/instrumentation/automatic-instrumentation.mdx (lines 270, 280, 347-348, 358-359)

philprime commented on docs/platforms/apple/common/migration/index.mdx:
```diff
@@ -3,7 +3,126 @@ title: Migration Guide
+**Removed deprecated profiling options from `SentryOptions`:**
+- `profilesSampleRate` - used for transaction-based profiling (deprecated)
+- `profilesSampler` - used for transaction-based profiling (deprecated)
+- `enableAppLaunchProfiling` - used for launch profiling (deprecated)
```
We should also mark these as removed in v9 in `docs/platforms/apple/common/profiling/index.mdx`

philprime reviewed (commented):
Thank you for preparing this, I found a couple of improvement opportunities we should adapt in this PR or in a follow-up.

I used this prompt to cross-check with the `CHANGELOG` in sentry-cocoa:

```
Use `gh` to view the PR description and changes of the current branch's PR. We want to have a migration guide for the sentry-cocoa v9 release. You can find the changelog for v9 at /path/to/sentry-cocoa/CHANGELOG.md

Analyze the proposed changes and give me a list of changes. A good list entry looks like this:

---
Developers upgrading to v9.0.0 will encounter contradictory documentation regarding `enablePerformanceV2`. The new migration guide explicitly states that `enablePerformanceV2` was removed, but the `automatic-instrumentation.mdx` file, which was not updated in this pull request, still contains instructions and code examples for enabling this option. This inconsistency will lead developers to attempt to configure a non-existent option, causing confusion and potential build or runtime errors.

**Suggested Fix:**

Update `docs/platforms/apple/common/tracing/instrumentation/automatic-instrumentation.mdx` to remove all references to `enablePerformanceV2` and clarify that Performance V2 is now enabled by default.
---

IMPORTANT: Keep in mind that the docs are used by developers/humans and should be concise and easy to understand. For each option and API we offer, check the documentation to find inconsistencies
```

philipphofmann commented:
I also used Claude to add the modified code snippets to our sample apps to validate that they're compiling 

<details><summary>Claude Test Run</summary>
<p>

# Snippet Compilation Report for Sentry Cocoa v9 Migration Guide

- **Repository:** `getsentry/sentry-cocoa`
- **Purpose:** Demonstrate that every Swift and Objective-C snippet added or changed in [feat(apple): Migration Guide for V9](https://github.com/getsentry/sentry-docs/pull/15649) compiles against the SDK.
- **Method:** Embed each snippet verbatim in debug-only helpers inside the sample apps so the compiler type-checks them without altering runtime behavior. Remove any obsolete (v8) variants that no longer build.

---

## Updated Files in the Cocoa Repo

1. `Samples/SentrySampleShared/SentrySampleShared/SentrySDKWrapper.swift`
   - Added `#if DEBUG private func migrationGuideV9SwiftDocSnippets()`.
   - Contains five guarded `if false` blocks, each wrapping a Swift initializer identical to the docs.
   - Helper body:

     ```swift
     #if DEBUG
     private func migrationGuideV9SwiftDocSnippets() {
         if false {
             SentrySDK.start { options in
                 options.dsn = "___PUBLIC_DSN___"
                 options.configureProfiling = {
                     $0.sessionSampleRate = 1
                     $0.lifecycle = .manual // or .trace for trace-based lifecycle
                 }
             }
         }

         if false {
             SentrySDK.start { options in
                 options.dsn = "___PUBLIC_DSN___"
                 options.enableFileIOTracing = true
                 options.enableFileManagerSwizzling = true
             }
         }

         if false {
             SentrySDK.start { options in
                 options.dsn = "___PUBLIC_DSN___"
                 options.enableFileIOTracing = true
                 options.enableDataSwizzling = false
                 options.enableFileManagerSwizzling = false
             }
         }

         if false {
             SentrySDK.start { options in
                 options.dsn = "___PUBLIC_DSN___"
                 options.enableLogs = true

                 options.beforeSendLog = { log in
                     if log.level == .info {
                         return nil
                     }
                     return log
                 }
             }
         }

         if false {
             SentrySDK.start { options in
                 options.dsn = "___PUBLIC_DSN___"
                 options.enableLogs = true
             }
         }
     }
     #endif
     ```

2. `Samples/iOS-ObjectiveC/iOS-ObjectiveC/AppDelegate.m`
   - Added `#if DEBUG - (void)docs_migrationGuideV9ObjectiveCSnippetsForCompiler`.
  - Contains five guarded `if (NO)` blocks mirroring the Objective-C versions.
  - Helper body:

    ```objc
    #if DEBUG
    - (void)docs_migrationGuideV9ObjectiveCSnippetsForCompiler
    {
        if (NO) {
            [SentrySDK startWithConfigureOptions:^(SentryOptions *options) {
                options.dsn = @"___PUBLIC_DSN___";
                options.configureProfiling = ^(SentryProfileOptions *_Nonnull profiling) {
                    profiling.sessionSampleRate = 1;
                    profiling.lifecycle = SentryProfileLifecycleManual; // or SentryProfileLifecycleTrace
                };
            }];

            [SentrySDK startWithConfigureOptions:^(SentryOptions *options) {
                options.dsn = @"___PUBLIC_DSN___";
                options.enableFileIOTracing = YES;
                options.enableFileManagerSwizzling = YES;
            }];

            [SentrySDK startWithConfigureOptions:^(SentryOptions *options) {
                options.dsn = @"___PUBLIC_DSN___";
                options.enableFileIOTracing = YES;
                options.enableDataSwizzling = NO;
                options.enableFileManagerSwizzling = NO;
            }];

            [SentrySDK startWithConfigureOptions:^(SentryOptions *options) {
                options.dsn = @"___PUBLIC_DSN___";
                options.enableLogs = YES;

                options.beforeSendLog = ^SentryLog *_Nullable(SentryLog *log) {
                    if (log.level == SentryLogLevelInfo) {
                        return nil;
                    }
                    return log;
                };
            }];

            [SentrySDK startWithConfigureOptions:^(SentryOptions *options) {
                options.dsn = @"___PUBLIC_DSN___";
                options.enableLogs = YES;
            }];
        }
    }
    #endif
    ```

Each block configures `SentrySDK.start` exactly as in the documentation. Because the functions live in production targets (wrapped in debug-only guards), Xcode validates that the APIs still exist and the snippets compile, while the guards prevent the code from executing.

---

## Snippets Verified (Swift)

All located inside `migrationGuideV9SwiftDocSnippets()`:

1. Configure UI profiling via `options.configureProfiling`.
2. Enable File I/O tracing with `enableFileManagerSwizzling`.
3. Disable `enableDataSwizzling` and `enableFileManagerSwizzling` for manual tracing.
4. Enable structured logging and filter info logs via `beforeSendLog`.
5. Enable logging without filters.

---

## Snippets Verified (Objective-C)

All located inside `docs_migrationGuideV9ObjectiveCSnippetsForCompiler`:

1. Configure UI profiling via `options.configureProfiling`.
2. Enable File I/O tracing plus FileManager swizzling.
3. Disable `enableDataSwizzling`/`enableFileManagerSwizzling`.
4. Enable structured logging with a `beforeSendLog` filter.
5. Enable logging without filters.

---

## Known Doc Discrepancy

The documentation snippet for Objective-C FileManager swizzling uses `options.dsn = "..."`; the compile-checked version in the repo corrects this to `@"..."`. Consider updating the doc PR accordingly.

---

## How to Repeat This Verification

1. **Identify snippets**
   - Fetch the PR’s modified files with  
     `curl -k -s https://api.github.com/repos/getsentry/sentry-docs/pulls/<PR>/files`  
     to list affected docs.
   - Download the raw content for each relevant doc to copy the snippets.

2. **Update helper functions**
   - Edit `Samples/SentrySampleShared/SentrySDKShared/SentrySDKWrapper.swift` and `Samples/iOS-ObjectiveC/iOS-ObjectiveC/AppDelegate.m`.
   - Replace the guarded blocks inside the existing debug helpers with the latest snippets (`if false` / `if (NO)` guards keep them from running).

3. **Remove obsolete code**
   - Delete snippet blocks referencing deprecated APIs (for example `options.experimental...`).

4. **Compile check**
   - Build the sample targets (or rely on CI). Because the snippets live in production sources, the compiler will flag API mismatches automatically.

5. **Document the run**
   - Note any adjustments (such as syntax fixes) and the files touched, then update this report.

This workflow makes it straightforward to refresh the helpers whenever the migration guide changes, ensuring the published examples continue to compile.

---

[^1]: [feat(apple): Migration Guide for V9](https://github.com/getsentry/sentry-docs/pull/15649)



</p>
</details>
